clippy_lints/src/methods/iter_on_single_or_empty_collections.rs

```diff
@@ -25,7 +29,25 @@ impl IterType {
     }
 }
 
-pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, method_name: &str, recv: &Expr<'_>) {
+fn is_arg_ty_unified_in_fn<'tcx>(
+    cx: &LateContext<'tcx>,
+    fn_id: DefId,
+    arg_id: HirId,
+    args: impl IntoIterator<Item = &'tcx Expr<'tcx>>,
+) -> bool {
+    let fn_sig = cx.tcx.fn_sig(fn_id).instantiate_identity();
+    let arg_id_in_args = args.into_iter().position(|e| e.hir_id == arg_id).unwrap();
+    let arg_ty_in_args = fn_sig.input(arg_id_in_args);
+
+    cx.tcx.predicates_of(fn_id).predicates.iter().any(|(clause, _)| {
```

Comment:
Maybe an interesting followup (in addition to looking through predicates, doesn't need to be done in this PR) would be to run a type visitor through each parameter type and check if the argument type is in there. I.e. when we have `fn<T>(&mut T, T)` and the argument maps to `T`, figure out that it unifies with the `&mut T` parameter. That would catch both of the false positives I mentioned before (mem::replace & unwrap_or).

Reply:
Added that as well & added a use of `std::mem::replace` to the tests, seems to be fine, not linted